# Microdroplet screening rapidly profiles a biocatalyst to enable its AI-assisted engineering

[![bioRxiv](https://img.shields.io/badge/bioRxiv-10.1101/2024.04.08.588565-b31b1b.svg)](https://www.biorxiv.org/content/10.1101/2024.04.08.588565)
[![Devices](https://img.shields.io/badge/%F0%9F%92%A7%20OpenWetWare-DropBase%20lrDMS%20chips-blue?style=flat&labelColor=gray&color=lightblue&link=https%3A%2F%2Fopenwetware.org%2Fwiki%2FDropBase%3ADevices)](https://openwetware.org/wiki/DropBase:Devices)
[![Data](https://zenodo.org/badge/doi/10.5281/zenodo.5123296.svg)](TODO)  
[![Published paper](https://img.shields.io/badge/%F0%9F%93%83_full_paper-pdf-green)](https://www.biorxiv.org/content/10.1101/2024.04.08.588565.full.pdf)
<a target="_blank" href="https://colab.research.google.com/drive/16rXKgbGXBBsHvS_2V84WbfKsJYf9lO4Q">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

This repo contains the processing scripts, code and evaluation methods for the paper "[Microdroplet screening rapidly profiles a biocatalyst to enable its AI-assisted engineering](TODO)".

## üì¶ Installation

**Via github**:

```bash
git clone https://github.com/Hollfelder-Lab/lrDMS-IRED.git
cd lrDMS-IRED
pip install .
```

## üöÄ Usage

### UMIC-seq2 suite for analysis of Oxford Nanopore data

**Extraction of UMIs from basecalled Nanopore reads for clustering**
Use the following commands to run the script 'UMI_extract.py': 
``` 
-T Number of threads to execute in parallel. Defaults to CPU count.  
-v Gives Version
-i Provide basecalled reads in fastq format.  
-o Specify the name of the UMI output fasta file.  
--probe A short sequence (eg 50 bp) adjacent to the UMI in fasta format.  
--umi_loc Location of UMI in reference to the probe. Upstream (up) or downstream (down).  
--umi_len Length of the UMI to be extracted.  
--min_probe_score Minimal alignment score of probe for processing. Defaults to length of probe sequence.  
```

**Using the mmseqs2 output to generate a .fasta files for individual clusters**  
Use the following commands to run the script 'tsv2fasta.py':  
Positional argument: 'clusterfile' Provide MMseqs2 output file (as .tsv)  
--sizethreshs Low and high thresholds for cluster sizes. Defaults to 1 1000.  
--splitfile If a read file (fasta/fastq) is provided here, individual fasta files per cluster will be generated.  
--similarityfile If a read file (fasta/fastq) is provided here, <similaritysamplesize> clusters will be analysed for internal cluster similarity.  
--similaritysamplesize Sample size for within cluster similarity analysis. Defaults to 500.  

**An example for a python script generating an .sh file that can be used to automate consensus generation 'clustering_automator.py'**
To automate consensus generation via medaka for clusterfiles generated using the 'tsv2fasta.py' script the 'clustering_automator.py' script can be used to generate a .sh file that to iteratively call all tools neccesary for consensus generation (minimap2, racon, mini_align, medaka consensus, medaka variant and the vcf2fasta python script.

**'vcf2fasta.py' for generating a .fasta file from the vcf file generated by medaka variant**
Can be used in automation by using the 'clustering_automator.py' script

**Generating a variant identifyer file for DiMSum**
Use the following arguments to run 'VIF_generator.py':  
--probe_UMI_down 'A short sequence (eg 14 bp) beginning 4 bp downstream of the UMI in fasta format.'  
--probe_UMI_up 'A short sequence (eg 14 bp) ending 4 bp upstream of the UMI in fasta format.'  
--probe_frame_down 'A short sequence (eg 10 bp) flanking the frame (downstream) in fasta format.  
--probe_frame_up '(eg 10 bp) flanking the frame (upstream) in fasta format.  
--clusterfolder 'path to clusterfolder generated after using clustering_automator.py'  

### Epistasis Graph Analysis

The easiest way to run the scripts is via Google Colab. The Jupyter notebook Epistasis_analysis.ipynb can directly be opened on Colab by pressing the following Open in Colab button:

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Hollfelder-Lab/lrDMS-IRED/blob/main/notebooks/Epistasis_analysis.ipynb)

After having opened the notebook in Colab, the python scripts `analysis_utils.py` and `plotting_utils.py` from the src folder need to be uploaded. These scripts contain the core functions for the analyses and plots to be carried out. Also, the csv file `MePy_merge_filtered_260923.csv` from the data folder containing the DMS data along side the distance matrix npy file `min_dimer_distances.npy` must be uploaded

## üß™ Data

Our data is available on Zenodo: [![DOI](https://zenodo.org/badge/doi/TODO/zenodo.TODO.svg)](TODO)

## üìú License

## üìÉ Citing this work

Please cite our paper if you use this code or data in your own work:

```bibtex
@article {
 Gantz2024,
 author = {Gantz, Maximilian and Mathis, Simon V. and Nintzel, Friederike E. H. and Zurek, Paul J. and Knaus, Tanja and Patel, Elie and Boros, Daniel and Weberling, Friedrich-Maximilian and Kenneth, Matthew R. A. and   Klein, Oskar J. and Medcalf, Elliot J. and Moss, Jacob and Herger, Michael and Kaminski, Tomasz S. and Mutti, Francesco G. and Lio, Pietro and Hollfelder, Florian},
 title = {Microdroplet screening rapidly profiles a biocatalyst to enable its AI-assisted engineering},
 elocation-id = {2024.04.08.588565},
 year = {2024},
 doi = {10.1101/2024.04.08.588565},
 publisher = {Cold Spring Harbor Laboratory},
 URL = {https://www.biorxiv.org/content/early/2024/04/08/2024.04.08.588565},
 eprint = {https://www.biorxiv.org/content/early/2024/04/08/2024.04.08.588565.full.pdf},
 journal = {bioRxiv}
}
```

## üë• Authors

- [Hollfelder Lab](https://hollfelder.bioc.cam.ac.uk/), Department of Biochemistry, University of Cambridge, UK
- [Lio Lab](https://www.cst.cam.ac.uk/people/pl219), Department of Computer Science and Technology, University of Cambridge, UK

## üìß Contact

For questions, please contact

- fh111(at)cam.ac.uk  
- mg985(at)cam.ac.uk
- simon.mathis(at)cl.cam.ac.uk
- fmw37(at)cam.ac.uk

## ü§ù Contributing

We welcome contributions to this repository. To set up the development environment, please follow the instructions below:

```bash
git clone https://github.com/Hollfelder-Lab/lrDMS-IRED.git
cd lrDMS-IRED
chmod +x contribute.sh
./contribute.sh
```
